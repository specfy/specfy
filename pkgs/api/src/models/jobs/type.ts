import type { Prisma } from '@prisma/client';

export interface JobDeployConfig {
  url: string;
  autoLayout?: boolean;
  hook?: {
    id: string;
    ref: string;
  };
}

export type JobWithOrgProject = Prisma.JobsGetPayload<{
  include: { Org: true; Project: true };
}>;

export type JobCode =
  | 'cant_auth_github'
  | 'failed_to_checkout'
  | 'failed_to_cleanup'
  | 'failed_to_clone'
  | 'failed_to_deploy'
  | 'failed_to_start_github_deployment'
  | 'failed_to_teardown'
  | 'missing_dependencies'
  | 'no_api_key'
  | 'org_not_installed'
  | 'project_not_installed'
  | 'success'
  | 'unknown';
