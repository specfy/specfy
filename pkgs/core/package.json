{
  "name": "@specfy/core",
  "version": "0.0.1",
  "description": "",
  "main": "build/index.js",
  "typings": "build/index.d.ts",
  "module": "build/index.js",
  "type": "module",
  "scripts": {
    "test": "vitest",
    "test:ci": "vitest run",
    "build": "tsc -p tsconfig.json"
  },
  "files": [
    "build"
  ],
  "author": "",
  "license": "MIT",
  "dependencies": {
    "@sentry/node": "7.66.0",
    "datadog-metrics": "0.11.0",
    "env-schema": "5.2.0",
    "logsnag": "1.0.0",
    "nanoid": "4.0.2",
    "pino": "8.15.1",
    "pino-pretty": "10.2.0",
    "slugify": "1.6.5",
    "zod": "3.22.3"
  },
  "devDependencies": {
    "typescript": "5.1.6"
  }
}
