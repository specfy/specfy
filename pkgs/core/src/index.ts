export * from './api.js';
export * from './avatar.js';
export * from './env.js';
export * from './diff.js';
export * from './id.js';
export * from './logger.js';
export * from './metric.js';
export * from './object.js';
export * from './sentry.js';
export * from './string.js';
export * from './types.utils.js';
export * from './validators/index.js';
export * from './types/github.js';
