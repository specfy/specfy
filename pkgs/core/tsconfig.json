{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./build",
    "declaration": true
  },
  "include": [
    "src",
  ]
}
