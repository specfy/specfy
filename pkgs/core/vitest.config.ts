// eslint-disable-next-line import/no-unresolved
import { defineConfig } from 'vitest/config';

export default defineConfig({
  test: {
    environment: 'node',
    include: ['src/**/*.test.ts'],
    clearMocks: true,
    coverage: {
      enabled: true,
      reporter: ['cobertura'],
      provider: 'v8',
    },
  },
  json: {},
});
