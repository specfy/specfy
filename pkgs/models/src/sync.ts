// TODO: find a better spot
export interface SyncConfig {
  orgId: string;
  projectId: string;
  branch: string;
  documentation?: SyncConfigDocumentation;
  stack?: SyncConfigStack;
}
export interface SyncConfigDocumentation {
  enabled: boolean;
  path?: string;
}
export interface SyncConfigStack {
  enabled: boolean;
  path?: string;
}
export type SyncConfigFull = {
  documentation: Required<SyncConfigDocumentation>;
  stack: Required<SyncConfigStack>;
};

export const def: SyncConfig = {
  orgId: '',
  projectId: '',
  branch: 'main',
  documentation: {
    enabled: true,
    path: 'docs/',
  },
  stack: {
    enabled: true,
    path: '/',
  },
};
