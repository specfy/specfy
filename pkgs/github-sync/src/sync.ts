import fs from 'node:fs/promises';
import path from 'node:path';

import type { Logger } from '@specfy/core';
import { l as defaultLogger } from '@specfy/core';
import type { JobMark } from '@specfy/models';
import type { Payload } from '@specfy/stack-analyser';
import {
  flatten,
  analyser,
  FSProvider as StackProvider,
  rules,
} from '@specfy/stack-analyser';
import figures from 'figures';

import { checkNothingMsg, checkPaths } from './common/helper.js';
import { listing } from './listing/index.js';
import type { ProviderFile } from './provider/base.js';
import { FSProvider } from './provider/fs.js';
import type { TransformedFile } from './transform/index.js';
import { transform } from './transform/index.js';
import {
  closeOldRevisions,
  getRevision,
  merge,
  prepBody,
  upload,
} from './upload/index.js';

// eslint-disable-next-line import/extensions
import '@specfy/stack-analyser/dist/rules/index.js';
rules.loadAll();

export interface SyncOptions {
  root: string;
  token: string;
  orgId: string;
  projectId: string;
  docEnabled: boolean;
  docPath: string;
  stackEnabled: boolean;
  stackPath: string;
  autoLayout: boolean;
  hostname?: string;
  logger?: Logger;
}

export class ErrorSync extends Error {
  code: JobMark['code'];
  constructor(code: JobMark['code']) {
    super(code);
    this.code = code;
  }
}

export async function sync({
  root,
  token,
  orgId,
  projectId,
  docEnabled,
  docPath,
  stackEnabled,
  stackPath,
  autoLayout,
  hostname = 'https://api.specfy.io',
  logger = defaultLogger,
}: SyncOptions) {
  const paths = await checkPaths({
    docEnabled,
    docPath,
    stackEnabled,
    stackPath,
    logger,
  });
  if (!paths) {
    throw new ErrorSync('sync_invalid_path');
  }

  const l = logger;
  const baseUrl = `${hostname}/0`;

  l.info('');

  // ------- Documentation
  let docs: TransformedFile[] = [];
  if (docEnabled) {
    l.info('-- Documentation');
    l.info(`${figures.triangleRight} Syncing ${docPath}`);

    l.info('Listing...');
    // --- List
    const list: ProviderFile[] = [];
    const provider = new FSProvider({
      path: docPath,
      ignorePaths: [],
    });

    await listing(
      {
        provider,
        acc: list,
      },
      '/'
    );
    l.info(`Listed ${figures.tick}`);

    // --- Transform
    l.info('Parsing...');
    docs = await transform(provider, list);
    l.info(`Parsed ${figures.tick}`);

    // Log
    l.info(`${figures.arrowRight} Found ${list.length} files`, {
      files: list.map((file) => file.fp),
    });
    l.info('');
  } else {
    l.warn(`${figures.info} Documentation Skipped`);
  }

  // ------- Stack
  let stack: Payload | null = null;
  if (stackEnabled) {
    l.info('');
    l.info('-- Stack');

    l.info(`${figures.triangleRight} Syncing: ${stackPath}`);

    l.info(`Analyzing...`);
    stack = flatten(
      await analyser({
        provider: new StackProvider({
          path: stackPath,
          ignorePaths: [],
        }),
      })
    );
    l.info(`Analyzed ${figures.tick}`);

    // output to folder for debug / manual review
    const file = path.join(root, 'stack.json');
    await fs.writeFile(file, JSON.stringify(stack.toJson(root), undefined, 2));
    // l.info(`${figures.arrowRight} Output`);
    // l.info('Stack', stack.toJson(root));
  } else {
    l.warn(`${figures.info} Stack Skipped`);
  }

  if (!docEnabled && !stackEnabled) {
    checkNothingMsg(logger);
    return;
  }

  // --- Upload
  l.info('');
  l.info('-- Deploy');
  l.info('Uploading to Specfy...');
  const body = prepBody({
    orgId,
    projectId,
    docs,
    stack,
    autoLayout,
    baseUrl,
    root,
  });

  const resUp = await upload({ body, token, baseUrl, logger });
  if ('error' in resUp) {
    // There is nothing to commit
    if (
      resUp.error.code === 'cant_create' &&
      resUp.error.reason === 'no_diff'
    ) {
      l.info(`Uploaded ${figures.tick}`);
      l.info(`${figures.info} No diff with production`);

      // It's perfectly okay that there is no diff, could be that we have found nothing or just nothing has changed
      return;
    }

    l.error(`Uploaded ${figures.cross}`);
    if (resUp.error.code === 'validation_error') {
      for (const err of resUp.error.form) {
        l.error(`[${err.code}] ${err.message}`);
      }
      for (const [key, err] of Object.entries(resUp.error.fields)) {
        l.error(`[${err?.code}] ${err!.message} (in "${key}")`);
      }

      const form = resUp.error.form;
      if (form.length > 0 && form[0].code === 'too_many_documents') {
        throw new ErrorSync('too_many_documents');
      }
      throw new ErrorSync('failed_to_upload');
    }

    l.error(JSON.stringify(resUp.error, null, 2));
    throw new ErrorSync('failed_to_upload');
  }

  if (resUp.data.stats.stack && resUp.data.stats.stack?.deleted > 1) {
    l.error(resUp.data.stats, 'Fail Safe: too many deletion');
    throw new ErrorSync('fail_safe_too_many_deletion');
  }

  const id = resUp.data.id;

  const resGet = await getRevision({
    id,
    orgId,
    projectId,
    token,
    baseUrl,
    logger,
  });
  const get = await resGet.json();
  l.info(`Uploaded ${figures.tick}`);

  l.info('');
  l.info(`${figures.triangleRight} Revision created: ${get.data.url}`);

  // --- Deduplication
  l.info('');
  l.info('Closing old revisions');
  const count = await closeOldRevisions({
    id,
    orgId,
    projectId,
    token,
    baseUrl,
    logger,
  });
  l.info(`Closed (${count}) ${figures.tick}`);

  l.info(`Merging`);
  await merge({ id, orgId, projectId, token, baseUrl, logger });
  l.info(`Merged ${figures.tick}`);
}
