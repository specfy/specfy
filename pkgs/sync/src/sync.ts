/* eslint-disable import/extensions */
import fs from 'node:fs/promises';
import path from 'node:path';

import { l as defaultLogger } from '@specfy/core';
import {
  flatten,
  analyser,
  FSProvider as StackProvider,
  rules,
} from '@specfy/stack-analyser';
import figures from 'figures';

import type { Logger } from '@specfy/core';
import type { CommitAnalysis, JobMark, SyncConfigFull } from '@specfy/models';
import type { Payload } from '@specfy/stack-analyser';

import { checkNothingMsg, checkPaths } from './common/helper.js';
import { analyzeCommit } from './git/index.js';
import { listing } from './listing/index.js';
import { FSProvider } from './provider/fs.js';
import { transform } from './transform/index.js';
import {
  closeOldRevisions,
  getRevision,
  merge,
  prepBody,
  upload,
} from './upload/index.js';

import type { ProviderFile } from './provider/base.js';
import type { TransformedFile } from './transform/index.js';

// eslint-disable-next-line import/order
import '@specfy/stack-analyser/dist/rules/index.js';
rules.loadAll();

export type SyncOptions = {
  root: string;
  token: string;
  orgId: string;
  projectId: string;
  sourceId: string;
  autoLayout: boolean;
  settings: SyncConfigFull;
  hostname?: string;
  logger?: Logger;

  // Internal
  jobId?: string | undefined;
};

export class ErrorSync extends Error {
  code: JobMark['code'];
  constructor(code: JobMark['code']) {
    super(code);
    this.code = code;
  }
}

export async function sync({
  root,
  token,
  orgId,
  projectId,
  sourceId,
  jobId,
  autoLayout,
  hostname = 'https://api.specfy.io',
  logger = defaultLogger,
  settings,
}: SyncOptions) {
  settings.documentation.path = path.join(root, settings.documentation.path);
  settings.stack.path = path.join(root, settings.stack.path);

  const paths = await checkPaths({ settings, logger });
  if (!paths) {
    throw new ErrorSync('sync_invalid_path');
  }

  const l = logger;
  const baseUrl = `${hostname}/0`;

  l.info('');

  // ------- Documentation
  let docs: TransformedFile[] = [];
  if (settings.documentation.enabled) {
    l.info('-- Documentation');
    l.info(`${figures.triangleRight} Syncing ${settings.documentation.path}`);

    l.info('Listing...');
    // --- List
    const list: ProviderFile[] = [];
    const provider = new FSProvider({
      path: settings.documentation.path,
      ignorePaths: [],
    });

    await listing(
      {
        provider,
        acc: list,
      },
      '/'
    );
    l.info(`Listed ${figures.tick}`);

    // --- Transform
    l.info('Parsing...');
    docs = await transform(provider, list);
    l.info(`Parsed ${figures.tick}`);

    // Log
    l.info(`${figures.arrowRight} Found ${list.length} files`, {
      files: list.map((file) => file.fp),
    });
    l.info('');
  } else {
    l.warn(`${figures.info} Documentation Skipped`);
  }

  // ------- Stack
  let stack: Payload | null = null;
  if (settings.stack.enabled) {
    l.info('');
    l.info('-- Stack');

    l.info(`${figures.triangleRight} Syncing: ${settings.stack.path}`);

    l.info(`Analyzing...`);
    stack = flatten(
      await analyser({
        provider: new StackProvider({
          path: settings.stack.path,
          ignorePaths: [],
        }),
      })
    );
    l.info(`Analyzed ${figures.tick}`);

    // output to folder for debug / manual review
    const file = path.join(root, 'stack.json');
    await fs.writeFile(file, JSON.stringify(stack.toJson(root), undefined, 2));
  } else {
    l.warn(`${figures.info} Stack Skipped`);
  }

  // ------- Git commit
  let commit: CommitAnalysis | null = null;
  if (settings.git.enabled) {
    l.info('');
    l.info('-- Git Commit Analysis');

    l.info(`Analyzing...`);
    commit = await analyzeCommit({ root });
    l.info(`Analyzed ${figures.tick}`);
  } else {
    l.warn(`${figures.info} Git Commit Skipped`);
  }

  if (!settings.documentation.enabled && !settings.stack.enabled) {
    checkNothingMsg(logger);
    return;
  }

  // --- Upload
  l.info('');
  l.info('-- Deploy');
  l.info('Uploading to Specfy...');
  const body = prepBody({
    orgId,
    projectId,
    sourceId,
    docs,
    stack,
    commit,
    autoLayout,
    baseUrl,
    root,
    jobId,
  });

  const resUp = await upload({ body, token, baseUrl, logger });
  if ('error' in resUp) {
    // There is nothing to commit
    if (
      resUp.error.code === 'cant_create' &&
      resUp.error.reason === 'no_diff'
    ) {
      l.info(`Uploaded ${figures.tick}`);
      l.info(`${figures.info} No diff with production`);

      // It's perfectly okay that there is no diff, could be that we have found nothing or just nothing has changed
      return;
    }

    l.error(`Uploaded ${figures.cross}`);
    if (resUp.error.code === 'validation_error') {
      for (const err of resUp.error.form) {
        l.error(`[${err.code}] ${err.message}`);
      }
      for (const [key, err] of Object.entries(resUp.error.fields)) {
        l.error(`[${err?.code}] ${err!.message} (in "${key}")`);
      }

      const form = resUp.error.form;
      if (form.length > 0 && form[0].code === 'too_many_documents') {
        throw new ErrorSync('too_many_documents');
      }
      throw new ErrorSync('failed_to_upload');
    }

    l.error(JSON.stringify(resUp.error, null, 2));
    throw new ErrorSync('failed_to_upload');
  }

  if (resUp.data.stats.stack && resUp.data.stats.stack?.deleted > 0) {
    l.error(resUp.data.stats, 'Fail Safe: too many deletion');
    throw new ErrorSync('fail_safe_too_many_deletion');
  }

  const id = resUp.data.id;

  const resGet = await getRevision({
    id,
    orgId,
    projectId,
    token,
    baseUrl,
    logger,
  });
  const get = await resGet.json();
  l.info(`Uploaded ${figures.tick}`);

  l.info('');
  l.info(`${figures.triangleRight} Revision created: ${get.data.url}`);

  // --- Deduplication
  l.info('');
  l.info('Closing old revisions');
  const count = await closeOldRevisions({
    id,
    orgId,
    projectId,
    token,
    baseUrl,
    logger,
  });
  l.info(`Closed (${count}) ${figures.tick}`);

  l.info(`Merging`);
  await merge({ id, orgId, projectId, token, baseUrl, logger });
  l.info(`Merged ${figures.tick}`);
}
