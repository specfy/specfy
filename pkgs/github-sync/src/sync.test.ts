import path from 'node:path';

import { dirname, type Logger } from '@specfy/core';
import { describe, expect, it } from 'vitest';

import type { SyncOptions } from './sync.js';
import { ErrorSync, sync } from './sync.js';

function getLogger() {
  const msgs: Array<Array<string | object>> = [];
  const log = (...args: string[]) => msgs.push(args);
  const logger: Logger = {
    info: log,
    error: log,
    warn: log,
    debug: log,
    fatal: log,
  } as unknown as Logger;
  return { msgs, logger };
}

function syncDefault(folder?: string): Omit<SyncOptions, 'logger'> {
  const root = folder || '/tmp/fdokfldkjlkgs';
  return {
    root,
    orgId: 'acme',
    projectId: 'dkjdkjf',
    token: '',
    autoLayout: true,
    docEnabled: true,
    docPath: `${root}/`,
    stackEnabled: true,
    stackPath: `${root}/`,
  };
}
describe('sync', () => {
  it('should stop if nothing is enabled', async () => {
    const { msgs, logger } = getLogger();
    const res = await sync({
      ...syncDefault(),
      logger,
      docEnabled: false,
      stackEnabled: false,
    });
    expect(res).toBeUndefined();
    expect(msgs).toMatchSnapshot();
  });

  it('should break if path does not exists', async () => {
    const { msgs, logger } = getLogger();
    await expect(() =>
      sync({
        ...syncDefault(),
        logger,
        docEnabled: true,
        stackEnabled: false,
      })
    ).rejects.toThrowError(new Error('sync_invalid_path'));
    expect(msgs).toMatchSnapshot();
  });

  it('should fail to upload', async () => {
    const { msgs, logger } = getLogger();
    const root = path.join(dirname, '../api/src/test/__fixtures__');
    await expect(() =>
      sync({
        ...syncDefault(root),
        logger,
        stackEnabled: false,
      })
    ).rejects.toThrowError(new ErrorSync('failed_to_upload'));

    // Clean snapshot
    msgs.forEach((row) =>
      row.forEach((msg, i) => {
        if (typeof msg === 'string') {
          row[i] = msg.replace(root, '');
        }
      })
    );
    expect(msgs).toMatchSnapshot();
  });
});
