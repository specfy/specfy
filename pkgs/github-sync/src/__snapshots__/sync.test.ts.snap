// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sync > should break if path does not exists 1`] = `
[
  [
    "✘ Path for documentation \\"/tmp/fdokfldkjlkgs/\\" does not exist",
  ],
]
`;

exports[`sync > should fail to upload 1`] = `
[
  [
    "",
  ],
  [
    "-- Documentation",
  ],
  [
    "▶ Syncing /Users/samuelbodin/code/specfy/pkgs/api/src/test/__fixtures__/",
  ],
  [
    "Listing...",
  ],
  [
    "Listed ✔",
  ],
  [
    "Parsing...",
  ],
  [
    "Parsed ✔",
  ],
  [
    "→ Found 5 files",
    {
      "files": [
        "/TestLink.md",
        "/TestLinkRelative.md",
        "/markdown-spec.md",
        "/nested/TestLinkNested.md",
        "/test-code.md",
      ],
    },
  ],
  [
    "",
  ],
  [
    "ℹ Stack Skipped",
  ],
  [
    "",
  ],
  [
    "-- Deploy",
  ],
  [
    "Uploading to Specfy...",
  ],
  [
    "POST https://api.specfy.io/0/revisions/upload",
  ],
  [
    "Uploaded ✘",
  ],
  [
    "{
  \\"code\\": \\"401_unauthorized\\"
}",
  ],
]
`;

exports[`sync > should stop if nothing is enabled 1`] = `
[
  [
    "",
  ],
  [
    "ℹ Documentation Skipped",
  ],
  [
    "ℹ Stack Skipped",
  ],
  [
    "",
  ],
  [
    "⚠ Nothing to do, did you mean to disable everything?",
  ],
]
`;
